8351851: Update PmemTest to run on AMD64

frkator opened the pull request:
Remove default OS platform linux x86_64, remove @ignores allowing amd64 as a viable OS platform selection

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8351851](https://bugs.openjdk.org/browse/JDK-8351851): Update PmemTest to run on AMD64 (**Task** - P3)


### Reviewers
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24416/head:pull/24416` \
`$ git checkout pull/24416`

Update a local copy of the PR: \
`$ git checkout pull/24416` \
`$ git pull https://git.openjdk.org/jdk.git pull/24416/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24416`

View PR using the GUI difftool: \
`$ git pr show -t 24416`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24416.diff">https://git.openjdk.org/jdk/pull/24416.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24416#issuecomment-2801958420)
</details>

mahendrachhipa commented:
LGTM

adinn reviewed (commented):
I'm not sure if this change is correct. The original directive was ignored and replaced with an x86_634 only directive because of automated tetsing concerns documented in [PR 12352](https://github.com/openjdk/jdk/pull/12352). If this change is happening because those automated test restrictions no longer apply for all three of amd64 (by  which I mean not intel x86_64) aarch64 and ppc then reverting to the ignored directive would be fine. If the only change is that amd64 now works as well as x86_64 then I would recommend modifying the first directive and keeping the ignored directive as the latter helps to clarify to general users how non-automated testing can be performed on aarch64 and ppc.

msheppar commented:
I would prefer if we retained the current test structure and swap
 the x86_64 with amd64, as amd64 is the os.arch on which the test will be run

Change 

 80 /* @test
 81  * @summary Testing NVRAM mapped byte buffer support
 82  * @run main/manual PmemTest
 83  * @requires (os.family == "linux")
 84  * @requires (os.arch == "x86_64")
 85  */
 86 
 87 /* @test
 88  * @summary Testing NVRAM mapped byte buffer support
 89  * @run main/manual PmemTest
 90  * @requires (os.family == "linux")
 91  * @requires ((os.arch == "amd64")|(os.arch == "aarch64")|(os.arch == "ppc64le"))
 92  * @ignore The test described here is currently disabled on systems that are not
 93  * x64-based and lack an external NVRAM memory device. In order to re-enable the
 94  * test, you will need to mount the NVRAM device, which will typically appear as
 95  * /dev/pmem0, to the directory /mnt/pmem. Once that is done, you can follow the
 96  * instructions above to create a test directory and remove the ignore tag.
 97  */

To  the preferred change i.e. to swap the x86_64 and amd64 and to use

80 /* @test
 81  * @summary Testing NVRAM mapped byte buffer support
 82  * @run main/manual PmemTest
 83  * @requires (os.family == "linux")
 84  * @requires (os.arch == “amd64”)
 85  */

 
 91  * @requires ((os.arch == “x86_64”)|(os.arch == "aarch64")|(os.arch == "ppc64le"))
 92  * @ignore The test described here is currently disabled on systems that are not

adinn commented:
> It would be preferable if we retained the current test structure and swap
the x86_64 with amd64, as amd64 is the os.arch on which the test will be run

When I wrote my previous comment I was under the impression that `@requires(os.arch == "x86_64")` selects Intel x86 64-bit only while `@requires(os.arch == "amd64")` selects either Intel x86 64-bit or AMD x86 64-bit. However, having looked through the code in files:
```
test/jdk/jdk/internal/util/ArchTest.java
src/java.base/share/classes/jdk/internal/util/Architecture.java
src/java.base/share/classes/jdk/internal/util/PlatformProps.java
/make/modules/java.base/gensrc/GensrcMisc.gmk
```
it appears to me that these two `@requires` clauses select the same thing i.e. either Intel x86 64-bit or AMD x86 64-bit. In which case the current config should mean that this test will run on AMD CPUs. Is that not the case?

msheppar commented:
> I'm not sure if this change is correct. The original directive was ignored and replaced with an x86_64-only directive because of automated testing concerns documented in [PR 12352](https://github.com/openjdk/jdk/pull/12352). If this change is happening because those automated test restrictions no longer apply for all three of amd64 (by which I mean not intel x86_64) aarch64 and ppc then reverting to the ignored directive would be fine. If the only change is that automated testing on amd64 now works as well as x86_64 then I would recommend modifying the first directive and keeping the ignored directive as the latter helps to clarify to general users how non-automated testing can be performed on aarch64 and ppc.

just to clarify that both instances  of @test  are **manual** and that current test environments are os.arch == "amd64", which based on the default  os.arch == "x86_64" meant the test was not run. As such, the purpose of the ticket is to make the default case "amd64"

msheppar commented:
> > It would be preferable if we retained the current test structure and swap
> > the x86_64 with amd64, as amd64 is the os.arch on which the test will be run
> 
> When I wrote my previous comment I was under the impression that `@requires(os.arch == "x86_64")` selects Intel x86 64-bit only while `@requires(os.arch == "amd64")` selects either Intel x86 64-bit or AMD x86 64-bit. However, having looked through the code in files:
> 
> ```
> test/jdk/jdk/internal/util/ArchTest.java
> src/java.base/share/classes/jdk/internal/util/Architecture.java
> src/java.base/share/classes/jdk/internal/util/PlatformProps.java
> /make/modules/java.base/gensrc/GensrcMisc.gmk
> ```
> 
> it appears to me that these two `@requires` clauses select the same thing i.e. either Intel x86 64-bit or AMD x86 64-bit. In which case the current config should mean that this test will run on AMD CPUs. Is that not the case?

Thanks for the details above. Currently the main linux execution platforms are returning "amd64" for the OS arch,  and that is not picked up by the current   @requires (os.arch == "x86_64")
so this is a pragmatic change to facilitate the manual execution of the most prevalent test environments.

It was observed that there might be a separate issue for jtreg on the @requires and
based on your feedback, and expectations there is a separate jtreg  issue to investigate the consistency of @requires on the os.arch attribute.

adinn commented:
@msheppar Thanks for clarifying what is at stake here. I am afraid don't know enough about the mechanics of the test infrastructure to explain why it is working the way you describe. However, given what you have observed I agree with your suggestion above that the right solution is to keep the two test declarations but switch "amd64" for "x86_64" and vice versa. That should make the tests work on amd64 CPUs while leaving it clear how anyone wanting to test on an alternative CPU would need to modify the test. I'll be happy to review the PR if it includes that change.

msheppar commented:
changing the @requires in the main test to
 @requires ((os.arch == "x86_64")|(os.arch == "amd64"))

seems reasonable 

and adding an id to each of the test cases is a good idea

frkator commented:
changes:
 - removed @ignores tag

and split to two test cases:
  - `@requires ((os.arch == "x86_64")|(os.arch == "amd64"))` with id `default_architecture_test_case`
  - `@requires ((os.arch == "aarch64")|(os.arch == "ppc64le"))` with id `other_architectures_test_case`
  
  local test shows that requires directive resolves properly as well as the test id is applied
  
    jtreg -version
    jtreg 7.6+1-12
    Installed in /home/isipka/apps/jtreg/jtreg-7.6+1/lib/jtreg.jar
    Running on platform version 25-ea from /home/isipka/.sdkman/candidates/java/25.ea.9-open.
    Built with Java(TM) 2 SDK, Version 11.0.16+11-LTS-199 on February 06, 2025.
    Copyright (c) 1999, 2022, Oracle and/or its affiliates. All rights reserved.
    Use is subject to license terms.
    JT Harness, version 6.0 ea b24 (September 27, 2022)
    Java Assembler Tools, version 7.0 ea b09 (June 04, 2020)
    TestNG: testng-7.3.0.jar, jcommander-1.82.jar, guice-5.1.0.jar
      JUnit: junit-platform-console-standalone-1.11.0.jar
      
      jtreg -verbose:esa java/nio/MappedByteBuffer/PmemTest.java
      failed to get value for jdk.static
      java.lang.UnsatisfiedLinkError: 'boolean jdk.test.whitebox.WhiteBox.isStatic()'
	      at jdk.test.whitebox.WhiteBox.isStatic(Native Method)
	      at requires.VMProps.isStatic(VMProps.java:830)
	      at requires.VMProps$SafeMap.put(VMProps.java:82)
	      at requires.VMProps.call(VMProps.java:145)
	      at requires.VMProps.call(VMProps.java:67)
	      at com.sun.javatest.regtest.agent.GetJDKProperties.run(GetJDKProperties.java:80)
	      at com.sun.javatest.regtest.agent.GetJDKProperties.main(GetJDKProperties.java:54)
      runner starting test: java/nio/MappedByteBuffer/PmemTest.java#default_architecture_test_case
      runner finished test: java/nio/MappedByteBuffer/PmemTest.java#default_architecture_test_case
      Failed. Execution failed: `main' threw exception: java.nio.file.NoSuchFileException: /mnt/pmem/test/pmemtest
      Test results: failed: 1; did not meet platform requirements: 1
      Report written to /home/isipka/workspace/jdk/open/test/jdk/JTreport/html/report.html
      Results written to /home/isipka/workspace/jdk/open/test/jdk/JTwork
      Error: Some tests failed or other problems occurred.

msheppar commented on test/jdk/java/nio/MappedByteBuffer/PmemTest.java:
```diff
@@ -77,23 +77,18 @@
  */
 
-/* @test
+/* @test id=other_architectures_test_case
```
we should to leave the @ignores for the 2nd test case 
the test will be run on amd64 environmenst , as such the addition of  | (os.arch == "amd64") is fine

frkator replied:
so just to confirm: we explitcitly want to ignore this test on `linux-aarch64` and `linux-ppc64le` regardless of is the required hardware present or not?

msheppar replied:
yes, retain the @ignore in the second test case.
The original aim of this issue is to change the default test case of os.arch == "x86_64" to os.arch == "amd64" as the test will by default be run in an amd64 environment.

frkator commented:
jtreg -ignore:quiet -verbose:esa java/nio/MappedByteBuffer/PmemTest.java
      failed to get value for jdk.static
      java.lang.UnsatisfiedLinkError: 'boolean jdk.test.whitebox.WhiteBox.isStatic()'
	      at jdk.test.whitebox.WhiteBox.isStatic(Native Method)
	      at requires.VMProps.isStatic(VMProps.java:830)
	      at requires.VMProps$SafeMap.put(VMProps.java:82)
	      at requires.VMProps.call(VMProps.java:145)
	      at requires.VMProps.call(VMProps.java:67)
	      at com.sun.javatest.regtest.agent.GetJDKProperties.run(GetJDKProperties.java:80)
	      at com.sun.javatest.regtest.agent.GetJDKProperties.main(GetJDKProperties.java:54)
      runner starting test: java/nio/MappedByteBuffer/PmemTest.java#default_architecture_test_case
      runner finished test: java/nio/MappedByteBuffer/PmemTest.java#default_architecture_test_case
      Failed. Execution failed: `main' threw exception: java.nio.file.NoSuchFileException: /mnt/pmem/test/pmemtest
      Test results: failed: 1; did not match keywords: 1
      Report written to /home/isipka/workspace/jdk/open/test/jdk/JTreport/html/report.html
      Results written to /home/isipka/workspace/jdk/open/test/jdk/JTwork
      Error: Some tests failed or other problems occurred.

frkator commented:
/integrate

adinn reviewed (approved):
Ok,that looks like a good solution

adinn commented:
@frkator Try to integrate again and I will sponsor.

frkator commented:
/integrate

adinn commented:
/sponsor
